File: sdk/node-sdk/src/ais.ts
/** AIS routing helpers for Node SDK clients. */

export function isAisSocketConfigured(env: Record<string, string | undefined> = process.env): boolean {
  return Boolean((env.CARACAL_AIS_UNIX_SOCKET_PATH ?? '').trim());
}

export function resolveAisBaseUrl(env: Record<string, string | undefined> = process.env): string | undefined {
  if (!isAisSocketConfigured(env)) {
    return undefined;
  }

  const host = (env.CARACAL_AIS_LISTEN_HOST ?? '127.0.0.1').trim() || '127.0.0.1';
  const port = (env.CARACAL_AIS_LISTEN_PORT ?? '7079').trim() || '7079';
  const rawPrefix = (env.CARACAL_AIS_API_PREFIX ?? '/v1/ais').trim() || '/v1/ais';
  const prefix = (rawPrefix.startsWith('/') ? rawPrefix : `/${rawPrefix}`).replace(/\/$/, '');

  return `http://${host}:${port}${prefix}`;
}

export function resolveSdkBaseUrl(env: Record<string, string | undefined> = process.env): string {
  return resolveAisBaseUrl(env) ?? env.CARACAL_API_URL ?? `http://localhost:${env.CARACAL_API_PORT ?? '8000'}`;
}

File: sdk/python-sdk/src/caracal_sdk/ais.py
"""AIS routing helpers for SDK clients."""

from __future__ import annotations

import os


def ais_socket_configured() -> bool:
    """Return True when an AIS Unix socket path is configured."""
    return bool((os.environ.get("CARACAL_AIS_UNIX_SOCKET_PATH") or "").strip())


def resolve_ais_base_url() -> str | None:
    """Resolve AIS base URL when socket-based AIS mode is enabled."""
    if not ais_socket_configured():
        return None

    host = (os.environ.get("CARACAL_AIS_LISTEN_HOST") or "127.0.0.1").strip() or "127.0.0.1"
    port = (os.environ.get("CARACAL_AIS_LISTEN_PORT") or "7079").strip() or "7079"
    prefix = (os.environ.get("CARACAL_AIS_API_PREFIX") or "/v1/ais").strip() or "/v1/ais"
    normalized_prefix = prefix if prefix.startswith("/") else f"/{prefix}"
    normalized_prefix = normalized_prefix.rstrip("/")

    return f"http://{host}:{port}{normalized_prefix}"


def resolve_sdk_base_url(default_port: str = "8000") -> str:
    """Resolve canonical SDK base URL with AIS routing preference."""
    ais_base = resolve_ais_base_url()
    if ais_base:
        return ais_base

    return os.environ.get("CARACAL_API_URL", f"http://localhost:{default_port}")
